### PR TITLE
[Processing] climb along line algorithm

### PR DESCRIPTION
**Pull request by ghtmtt**
## Description
Porting of the [Climb](https://plugins.qgis.org/plugins/Climb/) plugin of @havatv, kudos to him not to me.

Basically this algorithm takes a line and calculates the real ascending and descending made besides the minimum and maximum elevation.

Works if the line layer already has Z values, if not, the Drape algorithm can be used.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by luipir** [comment]
can you explain why this algorithm should be in core instead as provider in the original plugin? in this moment I'm +0 about it's merge.

**Comment by ghtmtt** [comment]
because I think that the analysis that it perform are useful for a wider audience and by having it in core allows to have more reliable tests ;)

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]: Is it ;)

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
            if field.name() == self.CLIMBATTRIBUTE:
```

Unrequired conversion?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]: Translate and possibly use `pushWarning` and remove prefix?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]: Translate

Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/Climb.py]: and possibly add information like feature id. In it's current state it's mostly spam ;)

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]: Better take a copy of the geometry first, `constParts()` directly works on the geometry parts so we need to make sure we keep local reference to a copy of the geometry in order to avoid suddenly having the base geometry modified or deleted (not sure how realistic that is, but taking a copy is on the safe side).

File: python/plugins/processing/algs/qgis/Climb.py
Comment (luipir) [inline python/plugins/processing/algs/qgis/Climb.py]: can flood log... becasue the message is too generic "Missing Z value", I would add a counter to avoid to repeat the same message (just in case)

Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/Climb.py]: Or collect all feature ids and just print a summary list once in the end

File: python/plugins/processing/algs/qgis/Climb.py
Comment (luipir) [inline python/plugins/processing/algs/qgis/Climb.py]: I sould add a column also with % repect part length not only delta Z.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (luipir) [reply python/plugins/processing/algs/qgis/Climb.py]: > Or collect all feature ids and just print a summary list once in the end

better

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/Climb.py]: Use formatting to make translation more flexible

`'{attr_name} attribute found and removed'.format(attr_name=self.CLIMBATTRIBUTE)`

**Comment by ghtmtt** [comment]
thanks @m-kuhn and @luipir for the review. I think travis will fail anyway because of the style. I cannot run `prepare_commit` script even if I compiled my local build with `astyle`. Any suggestion?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: done, thanks for the hint

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: Sorry, but I forgot this when adding two new attributes to the output:

         minelevindex = -1
         maxelevindex = -1

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: Sorry, but I have forgotten to include elifs for self.MINELEVATTRIBUTE and self.MAXELEVATTRIBUTE:

            elif field.name() == self.MINELEVATTRIBUTE:
                feedback.pushWarning(self.tr(
                                '{attr_name} attribute found and removed'.format(
                                    attr_name=self.MINELEVATTRIBUTE)
                                    )
                                )
                minelevindex = fieldnumber
            elif field.name() == self.MAXELEVATTRIBUTE:
                feedback.pushWarning(self.tr(
                                '{attr_name} attribute found and removed'.format(
                                    attr_name=self.MAXELEVATTRIBUTE)
                                    )
                                )
                maxelevindex = fieldnumber

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: Sorry again, but I forgot to add these new output attributes.

                if not (attrindex == climbindex or
                        attrindex == descentindex or
                        attrindex == minelevindex or
                        attrindex == maxelevindex):

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: I don't understand what you mean.  Could you explain it another way?

**Comment by m-kuhn** [comment]
Install autopep8

**Comment by ghtmtt** [comment]
@m-kuhn done `python3-autopep8` (astyle is already installed) but I always get this message when running `sh scripts/prepare-commit.sh`:

```
scripts/prepare-commit.sh: 29: scripts/prepare-commit.sh: [[: not found
astyle.sh not found
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: wait a second.. `pushWarning` is not a `feedback` method right? https://qgis.org/pyqgis/master/core/QgsProcessingFeedback.html

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: Since this reports the point number within a single line, it should be moved to line 226 to have a meaning.  The alternative is to include feature id and part number.
I also suggest to change the text to:
'The following points on the line do not have Z values: '

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: I think it it too much to report individual points that miss Z values, but if that is of interest we need to do something like this (and also add the partnumber variable):

    no_z_nodes.append('Feature: ' + str(feature.id() + ' part: ' + str(partnumber) + ' point: ' + str(idx))

Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/Climb.py]: Prefer a template based approach `'Feature: {feature_id}, Part: {part_id} ....'.format(feature.id(), partnumber)`, it will make translation easier (and code more readable).

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: Inherited for QgsFeedback.

Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/Climb.py]: oops, sorry for that. Then use `pushInfo` or `pushError`? but skip the `Warning:`, that's meta-information that shouldn't be embedded in the string (rather encoded as color / other formatting / ...).

**Comment by ghtmtt** [comment]
> Don't compile astyle but install it from your distribution (assuming you have a recent enough distribution to come with a modern version of astyle).

@m-kuhn I think I already have it (working on debian sid):
```

matteo@debian:~/lavori/QGIS/QGIS(processing_climb)$ aptitude search astyle
i   astyle                                                             - Source code indenter for C, C++, Objective-C, C#, and Java                   
p   libastyle-dev                                                      - Library header files for Artistic Style                                      
p   libastyle3                                                         - Shared library for Artistic Style                                            
p   libastylej-jni                                                     - Java JNI library for Artistic Style 
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: @havatv I think Matthias suggested to grab all the no Z nodes indexes and put them into a string to show at the end of the algorithm, else with a string with a lot of nodes it could be a little bit overkilling to push information each time. The approach of the next comment (with the enhancements to what the string should contain) is maybe better?

**Comment by m-kuhn** [comment]
@ghtmtt I think `autopep8` is the issue here (since it's about python files)

**Comment by ghtmtt** [comment]
> @ghtmtt I think autopep8 is the issue here (since it's about python files)

@m-kuhn even more weird:

```
matteo@debian:~/lavori/QGIS/QGIS(processing_climb)$ aptitude search autopep8
p   elpa-py-autopep8                                                   - use autopep8 to beautify a Python buffer                                     
i A python-autopep8                                                    - tool that automatically formats Python code to conform to PEP 8              
i   python3-autopep8                                                   - tool that automatically formats Python code to conform to PEP 8              
p   vim-autopep8                                                       - vim plugin to apply autopep8
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: partnumber = partnumber + 1

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]: partnumber = 0

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: Fine with me. I have added some suggestions to allow that.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (luipir) [reply python/plugins/processing/algs/qgis/Climb.py]: @havatv delta Z is not a useful information if not related how steep is related with length of the geometry part... e.g. 100m in 2km lenght is a different steep respect 100m in 100m lenght => https://en.wikipedia.org/wiki/Grade_(slope) . I would suggest to process also % not only climb. I agree that can be post processed later, but IMHO is something useful. The problem is that climb is the integration of all delta Z of the geometry parts and does not represent a real % but a medium %.

**Comment by m-kuhn** [comment]
@ghtmtt When running prepare-commit.sh it should print errors if these tools are missing. Caveat: prepare-commit.sh only checks uncommitted files. It doesn't help if you run it after the commit is already done.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [reply python/plugins/processing/algs/qgis/Climb.py]: Big +1 for template, much easier for translations indeed

**Comment by ghtmtt** [comment]
mmm nope.. I tried to run it before `git add`, after `git add`, after `git commit`. Always the same error. Im working on my forked repository on the local machine just on another branch.

**Comment by m-kuhn** [comment]
Run this line from the main QGIS source dir, what's the output?
```sh
echo "\n\n\n\n\n\n" >> python/plugins/processing/ProcessingPlugin.py && ./scripts/prepare-commit.sh
```

**Comment by ghtmtt** [comment]
@m-kuhn something seems working :) Got the problem, thanks!

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: @luipir, the purpose of this algorithm is to calculate the total / accumulated climb (and descent) for a line. So it is the delta Zs that provide useful information (the positive deltas contribute to climb, the negative deltas contribute to descent). An example of use is to calculate the total climb of a cycling race, a cross country racing ski track or a private training session.
I think that climb / descent as a percentage seldom would be of interest. A line will in general have some descending parts and some ascending parts, and I can't see what an average would be useful for (it would also be very easy to calculate from the numbers provided by this algorithm).
A thing that could be of intereste is the steepness/slope of the steepest ascent / descent.

**Comment by ghtmtt** [comment]
Travis is happy :) 
 @havatv I should have added all the changes and suggestions you made, same for @m-kuhn ones.

**Review by havatv — APPROVED** [review]

**Comment by havatv** [comment]
Should we remove "A QGIS plugin", as I think @m-kuhn suggests?

**Comment by ghtmtt** [comment]
done

**Comment by nyalldawson** [comment]
I'd also like to review this before merge please!

**Comment by DelazJ** [comment]
@ghtmtt there's a typo in the tag that would have created an issue report in docs...

**Comment by ghtmtt** [comment]
@DelazJ I think it is the title of the PR that generates the commit for the docs, isn't that? In other case, when squashing the commit it should be possible (not sure) to change the title of the commit. Else, do you have some suggestions?

**Comment by DelazJ** [comment]
@ghtmtt 
>I think it is the title of the PR that generates the commit for the docs, isn't that?

No. Not really actually. It's the merged commit (and its "subcommits") message or title that are used to generate the report (see https://github.com/qgis/QGIS-Sysadmin/blob/master/webhooks/github_feature_tracker.cgi#L427). So if at the merge time, the commit is of type `Merge pull request #10005 from ghtmtt/processing_climb` with no reference to `[feature]` or `[needs-docs]` in the description and you did not mention any in your commits (the current situation) then it fails to create the issue. And from experience, this often happens.

> In other case, when squashing the commit it should be possible (not sure) to change the title of the commit. Else, do you have some suggestions?

Yes, it's still possible for committers to rename at merging time (fingers crossed? 😄 ) or you can edit the title now and expect that it will be kept when merging.
Other suggestions? Rebase and edit the first commit? Squash all in one commit with the appropriate tag? The safest for me is anyway to have it in the commit you do yourself.
But given this discussion, I think you don't need to change anything anymore; the committer would take care of that 😉 .

**Comment by ghtmtt** [comment]
Thanks for the clarification @DelazJ. @nyalldawson wants do to some other review so I think I have to add some more commits. And then I can add `[feature][needs-doc]` in the text of on of these commits.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/Climb.py]: Can you switch to template based style here too

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
        feedback.pushInfo(self.tr('The following points to not have Z values: {no_z_report}'.format(
                                  no_z_report=str(no_z_nodes))
                                 )
                         )
```

**Comment by ghtmtt** [comment]
done the format string. I've put in the commit message `[feature][needs-doc]` just because the typo in the first commit (this should do the work @DelazJ )

File: python/plugins/processing/algs/qgis/Climb.py
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Climb.py]: On a semi-related note, make sure you return a dict from this method, or you'll get an exception

File: python/plugins/processing/algs/qgis/Climb.py
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Climb.py]: I'm unsure about this block -- elsewhere we create two lists of fields (existing and new), and then use QgsProcessingUtils::combineFields to ensure field have unique names when there's a clash. I think that same approach should be used here.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Climb.py]: Check for feature.hasGeometry(), and handle appropriately for null geometry features.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Climb.py]: This block would be much simplified if you adapt the earlier suggestion of using combineFields -- then you'd just append the new attributes directly to the end of the existing ones.

File: python/plugins/processing/algs/help/qgis.yaml
Comment (nyalldawson) [inline python/plugins/processing/algs/help/qgis.yaml]:
```suggestion
  Input layers must have Z values present. If Z values are not available, the "Drape" (set Z value from raster) algorithm may be used to add Z values from a DEM layer.
```

**Comment by nyalldawson** [comment]
Looking at this algorithm, it's treating each feature in isolation and it should be adapted to use QgsProcessingFeatureBasedAlgorithm to avoid the manual feature looping and sink handling.

**Comment by ghtmtt** [comment]
Thanks @nyalldawson for the review. I should have adapted all your suggestions. 

> Looking at this algorithm, it's treating each feature in isolation and it should be adapted to use QgsProcessingFeatureBasedAlgorithm to avoid the manual feature looping and sink handling

i wasn't able to find any `QgsProcessingFeatureBasedAlgorithm` (python) algorithm in the source repo so this is the only thing not adapted (yet). Is it mandatory? From what I'm reading a lot of algorithms should be adapted to this kind of class..

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: The problem with using combineFields is that the names of the climb, descent, ... fields will not be know in the output (could be "climb", "climb_1", "climb_23", ...).  This will make it difficult to use this algorithm in a chain of operations.  So I think the original solution is better in this case.

**Comment by havatv** [comment]
Regarding QgsProcessingFeatureBasedAlgorithm, the algorithm also has layer derived output (total climb for all lines, total descent, ...) in addition to the output features. Is that a problem?

**Comment by ghtmtt** [comment]
@nyalldawson should I do something here? Like restarting the build?

**Comment by luipir** [comment]
frestarted... btw the error is not related with this PR

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]: Check one PR from a few days ago
```suggestion
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
                self.tr('Line layer'),
```

AFAIK, we don't have parenthesis in processing labels

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
```

A `return` after a `raise` has no effect. It will raise a warning in a python editor.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
                no_geometry_report=no_geometry)
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
                no_z_report=no_z_nodes)
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
        out_fields = QgsProcessingUtils.combineFields(thefields, source_fields)
```

Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: This change will make sure that the names of the generated attributes will be known in the output ("colliding" names in the input will be renamed).

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
            attrs = []
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
           attr.extend(feature.attributes())

```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [inline python/plugins/processing/algs/qgis/Climb.py]:
```suggestion
            attrs.append(maxelev)
            attrs.extend(feature.attributes())
```

File: python/plugins/processing/algs/qgis/Climb.py
Comment (havatv) [reply python/plugins/processing/algs/qgis/Climb.py]: @nyalldawson, I have now suggested a solution where combineFields is still used, but with the new fields as the first parameter and the input fields as the second parameter.  Is that a solution you can accept?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: ?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: `no_geometry` is a list. AFAIK `feedback` only accepts strings, isn't it?

File: python/plugins/processing/algs/qgis/Climb.py
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: same of above

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [reply python/plugins/processing/algs/qgis/Climb.py]: https://github.com/qgis/QGIS/pull/10023

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [reply python/plugins/processing/algs/qgis/Climb.py]: So `', '.join(no_geometry)` ?

Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/Climb.py]: still don't get why changing this..

File: python/plugins/processing/algs/help/qgis.yaml
Comment (havatv) [inline python/plugins/processing/algs/help/qgis.yaml]:
```suggestion
  The output layer is a copy of the input layer with additional fields that contain the total climb, total descent, the minimum elevation and the maximum elevation for each line geometry. If the input layer contains fields with the same names as these added fields, they will be renamed (field names will be altered to "name_2", "name_3", etc, finding the first non-duplicate name).
```

**Comment by ghtmtt** [comment]
guys should I close the PR and reopen it?

**Comment by m-kuhn** [comment]
Because of feature freeze?
I don't think so, they normally stay open until the freeze is over and are merged subsequently if ready.

File: python/plugins/processing/algs/qgis/Climb.py
Comment (Gustry) [reply python/plugins/processing/algs/qgis/Climb.py]: Because it's prettier ;-)
```python
>>>no_geometry=['geom1', 'geom2']
>>>print(no_geometry)
['geom1', 'geom2']
>>>print(', '.join(no_geometry))
geom1, geom2
```
Anyway, no problem.

**Comment by ghtmtt** [comment]
>  I don't think so, they normally stay open until the freeze is over and are merged subsequently if ready.

+1, thanks

**Comment by Gustry** [comment]
Or you can add the `Merge after thaw` label, it helps stalebot for this situation ;-)

**Comment by m-kuhn** [comment]
Have all open questions been addressed?

**Comment by ghtmtt** [comment]
@m-kuhn I think so. I had another look and addressed @Gustry suggestions to split the list for the `feedback` method.

**Comment by m-kuhn** [comment]
Ah, sorry, it's hard to see with all the open comments. Can you close/collapse them (or does this need special access rights?)

**Pull request merged**
